File: fc-chat/demo/part3/ch3/04-message-system-add-channel/src/main/java/net/prostars/messagesystem/repository/UserConnectionRepository.java
package net.prostars.messagesystem.repository;

import java.util.Collection;
import java.util.List;
import java.util.Optional;
import net.prostars.messagesystem.constant.UserConnectionStatus;
import net.prostars.messagesystem.dto.projection.InviterUserIdProjection;
import net.prostars.messagesystem.dto.projection.UserConnectionStatusProjection;
import net.prostars.messagesystem.dto.projection.UserIdUsernameInviterUserIdProjection;
import net.prostars.messagesystem.entity.UserConnectionEntity;
import net.prostars.messagesystem.entity.UserConnectionId;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.lang.NonNull;
import org.springframework.stereotype.Repository;

@Repository
public interface UserConnectionRepository
    extends JpaRepository<UserConnectionEntity, UserConnectionId> {

  Optional<UserConnectionStatusProjection>
      findUserConnectionStatusByPartnerAUserIdAndPartnerBUserId(
          @NonNull Long partnerAUserId, @NonNull Long partnerBUserId);

  Optional<UserConnectionEntity> findByPartnerAUserIdAndPartnerBUserIdAndStatus(
      @NonNull Long partnerAUserId,
      @NonNull Long partnerBUserId,
      @NonNull UserConnectionStatus status);

  Optional<InviterUserIdProjection> findInviterUserIdByPartnerAUserIdAndPartnerBUserId(
      @NonNull Long partnerAUserId, @NonNull Long partnerBUserId);

  long countByPartnerAUserIdAndPartnerBUserIdInAndStatus(
      @NonNull Long partnerAUserId,
      @NonNull Collection<Long> partnerBUserIds,
      @NonNull UserConnectionStatus status);

  long countByPartnerBUserIdAndPartnerAUserIdInAndStatus(
          @NonNull Long partnerBUserId,
          @NonNull Collection<Long> partnerAUserIds,
          @NonNull UserConnectionStatus status);

  @Query(
      "SELECT u.partnerBUserId AS userId, userB.username AS username, u.inviterUserId AS inviterUserId "
          + "FROM UserConnectionEntity u "
          + "INNER JOIN UserEntity userB ON u.partnerBUserId = userB.userId "
          + "WHERE u.partnerAUserId = :userId AND u.status = :status")
  List<UserIdUsernameInviterUserIdProjection> findByPartnerAUserIdAndStatus(
      @NonNull @Param("userId") Long userId, @NonNull @Param("status") UserConnectionStatus status);

  @Query(
      "SELECT u.partnerAUserId AS userId, userA.username AS username, u.inviterUserId AS inviterUserId "
          + "FROM UserConnectionEntity u "
          + "INNER JOIN UserEntity userA ON u.partnerAUserId = userA.userId "
          + "WHERE u.partnerBUserId = :userId AND u.status = :status")
  List<UserIdUsernameInviterUserIdProjection> findByPartnerBUserIdAndStatus(
      @NonNull @Param("userId") Long userId, @NonNull @Param("status") UserConnectionStatus status);
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/repository/UserConnectionRepository.kt
package com.yscorp.withpush.messagesystem.repository

import com.yscorp.withpush.messagesystem.constant.UserConnectionStatus
import com.yscorp.withpush.messagesystem.dto.projection.InviterUserIdProjection
import com.yscorp.withpush.messagesystem.dto.projection.UserConnectionStatusProjection
import com.yscorp.withpush.messagesystem.dto.projection.UserIdUsernameInviterUserIdProjection
import com.yscorp.withpush.messagesystem.entity.UserConnectionEntity
import com.yscorp.withpush.messagesystem.entity.UserConnectionId
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository

@Repository
interface UserConnectionRepository : JpaRepository<UserConnectionEntity, UserConnectionId> {

    fun findUserConnectionStatusByPartnerAUserIdAndPartnerBUserId(
        partnerAUserId: Long,
        partnerBUserId: Long
    ): UserConnectionStatusProjection?

    fun findByPartnerAUserIdAndPartnerBUserIdAndStatus(
        partnerAUserId: Long,
        partnerBUserId: Long,
        status: UserConnectionStatus
    ): UserConnectionEntity?

    fun findInviterUserIdByPartnerAUserIdAndPartnerBUserId(
        partnerAUserId: Long,
        partnerBUserId: Long
    ): InviterUserIdProjection?

    fun countByPartnerAUserIdAndPartnerBUserIdInAndStatus(
        partnerAUserId: Long,
        partnerBUserIds: Collection<Long>,
        status: UserConnectionStatus?
    ): Long

    fun countByPartnerBUserIdAndPartnerAUserIdInAndStatus(
        partnerBUserId: Long,
        partnerAUserIds: Collection<Long>,
        status: UserConnectionStatus?
    ): Long

    @Query(
        """
        SELECT u.partnerBUserId AS userId, userB.username AS username, u.inviterUserId AS inviterUserId
        FROM UserConnectionEntity u
        INNER JOIN UserEntity userB ON u.partnerBUserId = userB.userId
        WHERE u.partnerAUserId = :userId AND u.status = :status
        """
    )
    fun findByPartnerAUserIdAndStatus(
        @Param("userId") userId: Long,
        @Param("status") status: UserConnectionStatus
    ): List<UserIdUsernameInviterUserIdProjection>

    @Query(
        """
        SELECT u.partnerAUserId AS userId, userA.username AS username, u.inviterUserId AS inviterUserId
        FROM UserConnectionEntity u
        INNER JOIN UserEntity userA ON u.partnerAUserId = userA.userId
        WHERE u.partnerBUserId = :userId AND u.status = :status
        """
    )
    fun findByPartnerBUserIdAndStatus(
        @Param("userId") userId: Long,
        @Param("status") status: UserConnectionStatus
    ): List<UserIdUsernameInviterUserIdProjection>
}
